privacy-deidentification/src/main/java/io/konig/privacy/deidentification/repo/PersonRepository.java

```diff
@@ -0,0 +1,164 @@
+import org.springframework.transaction.annotation.Transactional;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.databind.node.TextNode;
+import com.github.fge.jsonschema.core.exceptions.ProcessingException;
+
+import io.konig.privacy.deidentification.model.Identity;
+import io.konig.privacy.deidentification.model.Person;
+import io.konig.privacy.deidentification.model.PersonData;
+import io.konig.privacy.deidentification.model.PersonKeys;
+import io.konig.privacy.deidentification.model.PersonRowMapper;
+
+@Repository
+@Transactional
+public class PersonRepository {
+
+	final static SecureRandom secureRandom = new SecureRandom();
+
+	private static final String FETCH_QUERY = "SELECT DE_IDENTIFICATION.PERSON.PERSON_DATA, DE_IDENTIFICATION.PERSON.PSEUDONYM from "
+			+ "DE_IDENTIFICATION.PERSON  INNER JOIN DE_IDENTIFICATION.PERSON_EMAIL ON DE_IDENTIFICATION.PERSON.PSEUDONYM= DE_IDENTIFICATION.PERSON_EMAIL.PERSON_PSEUDONYM "
+			+ "AND trim(DE_IDENTIFICATION.PERSON_EMAIL.EMAIL_ID)=?";
+
+	@Autowired
+	JdbcTemplate template;
+
+	public List<PersonKeys> put(Person person, String version, String baseURL) throws ProcessingException, IOException {
+		ArrayNode tagsArray = (ArrayNode) person.getPerson().findValue("data");
+		String dataSourceId = person.getPerson().findValue("datasource").textValue();
+		List<PersonKeys> personKeyList = new ArrayList<PersonKeys>();
+		for (int i = 0; i < tagsArray.size(); i++) {
+			PersonKeys personKeys = new PersonKeys();
+			PersonData personData = null;
+			ObjectNode objectValue = (ObjectNode) tagsArray.get(i);
+			String pseudonym = null;
+			ArrayNode emailArray = (ArrayNode) objectValue.findValue("email");
+			List<String> email = new ArrayList<String>();
+			boolean IdentifierExists = false;
+			for (int j = 0; j < emailArray.size(); j++) {
+				TextNode textValue = (TextNode) emailArray.get(j);
+				IdentifierExists = identifierExists(textValue.textValue());
+				if (IdentifierExists) {
+
+					PersonRowMapper personRowMapper = new PersonRowMapper();
+					personData = (PersonData) template.queryForObject(FETCH_QUERY, personRowMapper,
+							textValue.textValue());
+
+					if (personData.getPerson() != null) {
+						ObjectMapper om = new ObjectMapper();
+						JsonNode beforeNode = om.readTree(personData.getPerson());
+						JsonNode afterNode = om.readTree(objectValue.toString());
+						JsonNode mergeNode = merge(beforeNode, afterNode);
+
+						String updateQuery = "UPDATE DE_IDENTIFICATION.PERSON SET PERSON_DATA=? WHERE PSEUDONYM=?";
+						template.update(updateQuery, mergeNode.toString(), personData.getPseudonym());
+
+						personKeys.setId(baseURL + personData.getPseudonym());
+						personKeys.setPseudonym(personData.getPseudonym());
+					}
+				} else {
+					pseudonym = randomString(30);
```

Comment:
This is a problem.  You could potentially be creating a new pseudonym multiple times in a loop (once for each unmatched email address).